File: javascript/064.js
/**
 * @param {number[][]} grid
 * @return {number}
 */
var minPathSum = function(grid) {
  const rowCount = grid.length;
  if (rowCount === 0) return 0;
  const colCount = grid[0].length;
  if (colCount === 0) return 0;

  for (let row = 0; row < rowCount; ++row) {
    for (let col = 0; col < colCount; ++col) {
      let currentSum = 0;
      if (row === 0 && col === 0) continue;
      if (row === 0) grid[row][col] = grid[row][col - 1] + grid[row][col];
      if (col === 0) grid[row][col] = grid[row - 1][col] + grid[row][col];
      if (row !== 0 && col !== 0) grid[row][col] = grid[row][col] + Math.min(grid[row - 1][col], grid[row][col - 1]);
    }
  }

  return grid[rowCount - 1][colCount - 1];
};

File: ruby/064.rb
# @param {Integer[][]} grid
# @return {Integer}
def min_path_sum(grid)
  row_count = grid.length
  col_count = grid[0].length

  0.upto(row_count - 1).each do |row|
    0.upto(col_count - 1).each do |col|
      next if row == 0 && col == 0

      if row == 0
        grid[row][col] += grid[row][col - 1]
        next
      end

      if col == 0
        grid[row][col] += grid[row - 1][col]
        next
      end

      up = grid[row - 1][col]
      left = grid[row][col - 1]
      grid[row][col] += (left < up ? left : up)
    end
  end

  grid[row_count - 1][col_count - 1]
end